src/chainstate/coordinator/mod.rs

```diff
@@ -682,6 +682,128 @@ fn forget_orphan_stacks_blocks(
     Ok(())
 }
 
+/// Get the heaviest affirmation map, when considering epochs.
+/// * In epoch 2.05 and prior, the heaviest AM was the sortition AM.
+/// * In epoch 2.1, the reward cycles prior to the 2.1 boundary remain the sortition AM.
+pub fn static_get_heaviest_affirmation_map(
+    burnchain: &Burnchain,
+    burnchain_blocks_db: &BurnchainDB,
+    sortition_db: &SortitionDB,
+    sortition_tip: &SortitionId,
+) -> Result<AffirmationMap, Error> {
+    let last_2_05_rc = sortition_db.get_last_epoch_2_05_reward_cycle()? as usize;
+
+    let sort_am = sortition_db.find_sortition_tip_affirmation_map(sortition_tip)?;
+
+    let heaviest_am = BurnchainDB::get_heaviest_anchor_block_affirmation_map(
+        burnchain_blocks_db.conn(),
+        burnchain,
+    )?;
+
+    let mut am_entries = vec![];
+    for i in 0..last_2_05_rc {
+        if i < sort_am.affirmations.len() {
+            am_entries.push(sort_am.affirmations[i]);
+        } else {
+            return Ok(AffirmationMap::new(am_entries));
+        }
+    }
+    for i in last_2_05_rc..heaviest_am.len() {
+        am_entries.push(heaviest_am.affirmations[i]);
+    }
+
+    Ok(AffirmationMap::new(am_entries))
+}
+
+/// Get the canonical affirmation map, when considering epochs.
+/// * In epoch 2.05 and prior, the heaviest AM was the sortition AM.
+/// * In epoch 2.1, the reward cycles prior to the 2.1 boundary remain the sortition AM.
+pub fn static_get_canonical_affirmation_map(
+    burnchain: &Burnchain,
+    burnchain_blocks_db: &BurnchainDB,
+    sortition_db: &SortitionDB,
+    chain_state_db: &StacksChainState,
+    sortition_tip: &SortitionId,
+) -> Result<AffirmationMap, Error> {
+    let last_2_05_rc = sortition_db.get_last_epoch_2_05_reward_cycle()? as usize;
+
+    let sort_am = sortition_db.find_sortition_tip_affirmation_map(sortition_tip)?;
+
+    let canonical_am = StacksChainState::find_canonical_affirmation_map(
```

Comment:
question: can we move the statement to line # 746? It looks like we may never use `canonical_am` if `last_2_05_rc` is bigger than `sort_am.affirmation.len()`.

Reply:
This isn't as tenable now that the affirmation map code here has been consolidated.  But, this method is called at most three times per sortition processed (it's usually only called twice, and _that_ could be reduced if we wanted), which is pretty rare in the system's operation.